o1vm/src/pickles/verifier.rs

```diff
@@ -0,0 +1,276 @@
+#![allow(clippy::type_complexity)]
+#![allow(clippy::boxed_local)]
+
+use ark_ec::{AffineRepr, Group};
+use ark_ff::{PrimeField, Zero /* One */};
+use rand::thread_rng;
+
+use kimchi::{
+    circuits::{
+        berkeley_columns::BerkeleyChallenges,
+        domains::EvaluationDomains,
+        expr::{ColumnEvaluations, Constants, Expr, ExprError, PolishToken},
+        gate::CurrOrNext,
+    },
+    curve::KimchiCurve,
+    groupmap::GroupMap,
+    plonk_sponge::FrSponge,
+    proof::PointEvaluations,
+};
+use mina_poseidon::{sponge::ScalarChallenge, FqSponge};
+use poly_commitment::{
+    commitment::{
+        absorb_commitment, combined_inner_product, BatchEvaluationProof, Evaluation, PolyComm,
+    },
+    ipa::OpeningProof,
+    OpenProof,
+};
+
+use super::{
+    column_env::get_all_columns,
+    proof::{Proof, WitnessColumns},
+};
+use crate::{interpreters::mips::column::N_MIPS_SEL_COLS, E};
+use kimchi_msm::columns::Column;
+
+type CommitmentColumns<G> = WitnessColumns<PolyComm<G>, [PolyComm<G>; N_MIPS_SEL_COLS]>;
+type EvaluationColumns<G> = WitnessColumns<
+    <<G as AffineRepr>::Group as Group>::ScalarField,
+    [<<G as AffineRepr>::Group as Group>::ScalarField; N_MIPS_SEL_COLS],
+>;
+
+// TODO: Move and perhaps derive some traits for these
+struct ColumnEval<'a, G: AffineRepr> {
+    commitment: &'a CommitmentColumns<G>,
+    zeta_eval: &'a EvaluationColumns<G>,
+    zeta_omega_eval: &'a EvaluationColumns<G>,
+}
+
+impl<G: AffineRepr> ColumnEvaluations<<G as AffineRepr>::ScalarField> for ColumnEval<'_, G> {
+    type Column = Column;
+    fn evaluate(
+        &self,
+        col: Self::Column,
+    ) -> Result<PointEvaluations<<G as AffineRepr>::ScalarField>, ExprError<Self::Column>> {
+        let ColumnEval {
+            commitment: _,
+            zeta_eval,
+            zeta_omega_eval,
+        } = self;
+        if let Some(&zeta) = zeta_eval.get_column(&col) {
+            if let Some(&zeta_omega) = zeta_omega_eval.get_column(&col) {
+                Ok(PointEvaluations { zeta, zeta_omega })
+            } else {
+                Err(ExprError::MissingEvaluation(col, CurrOrNext::Next))
+            }
+        } else {
+            Err(ExprError::MissingEvaluation(col, CurrOrNext::Curr))
+        }
+    }
+}
+
+pub fn verify<
+    G: KimchiCurve,
+    EFqSponge: Clone + FqSponge<G::BaseField, G, G::ScalarField>,
+    EFrSponge: FrSponge<G::ScalarField>,
+>(
+    domain: EvaluationDomains<G::ScalarField>,
```

Comment:
in a follow up we should replace the domain by omega and the domain size.
I do not like having something of linear size as the input of the verifier